lib: Add ofw from `plat/drivers` to `lib`

rares-miculescu opened the pull request:
### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Ofw (open firmware) driver is moved from `plat/drivers` into `lib/ukofw`.

razvanvirtan commented:
Hi @rares-miculescu, thanks for taking time to look into this!

Some first suggestions regarding the commits style.

First, the commits shouldn't reflect all the iterations of this PR, but only the final changes with respect to the current state of `staging`. In this case, I suggest having only 1 commit instead of 3. Basically, you need a commit that provides the changes between [unikraft:staging](https://github.com/unikraft/unikraft/tree/staging) and the current state of [rares-miculescu:move-ofw-test](https://github.com/rares-miculescu/unikraft/tree/move-ofw-test).

Also, make sure that the commit message follows the [unikraft rules]([rares-miculescu:move-ofw-test](https://github.com/rares-miculescu/unikraft/tree/move-ofw-test)) - I would use `lib/ukofw` instead of simple `lib` - and the general best practices for commit messages (useful reference [here](https://techblog.rosedu.org/git-good-practices.html), the `Commit Messages` section).

Let me know if you have any questions!

razvanvirtan reviewed (changes requested):
Hi @rares-miculescu, thanks for taking time to look into this!

Some first suggestions regarding the commits style.

First, the commits shouldn't reflect all the iterations of this PR, but only the final changes with respect to the current state of `staging`. In this case, I suggest having only 1 commit instead of 3. Basically, you need a commit that provides the changes between [unikraft:staging](https://github.com/unikraft/unikraft/tree/staging) and the current state of [rares-miculescu:move-ofw-test](https://github.com/rares-miculescu/unikraft/tree/move-ofw-test).

Also, make sure that the commit message follows the [unikraft rules](https://unikraft.org/docs/contributing/submitting-changes/)) - I would use `lib/ukofw` instead of simple `lib` - and the general best practices for commit messages (useful reference [here](https://techblog.rosedu.org/git-good-practices.html), the `Commit Messages` section).

Let me know if you have any questions!

StefanJum commented on lib/ukofw/Config.uk:
```diff
@@ -0,0 +1,3 @@
+menuconfig LIBUKOFW
+    bool "ukofw: Helper for libfdt"
```
Maybe add a `help` section in here?

StefanJum reviewed (changes requested):
Thank you @rares-miculescu. 

Besides the comments, please:
* [Squash](https://www.baeldung.com/ops/git-squash-commits) and [`sign-off`](https://docs.pi-hole.net/guides/github/how-to-signoff/) the commits
* Use `lib/ukofw:` as the commit selector/component.

razvanvirtan reviewed (approved):
@rares-miculescu I think you can move this PR to the `Ready for review` state. All good on my side :)

Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

michpappas commented on lib/ukofw/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBOFW
+       bool "ukofw: Open Firmware library support"
```
I know this description is moved verbatim from `plat/kvm/Config.uk` but it's quite inaccurate, so maybe it's a good opportunity to update it now. A better description would be "ukofw: Device tree helper functions".

michpappas commented on lib/ukofw/Makefile.uk:
```diff
@@ -0,0 +1,8 @@
+$(eval $(call addplatlib_s,kvm,libukofw,$(CONFIG_LIBUKOFW)))
```
This is no longer a platform library, so you should be using `addlib_s` instead:
```
$(eval $(call addlib_s,libukofw,$(CONFIG_LIBUKOFW)))
```

michpappas commented on lib/ukofw/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBOFW
```
This needs to be `LIBUKOFW`

michpappas commented on lib/ukofw/Makefile.uk:
```diff
@@ -0,0 +1,8 @@
+$(eval $(call addplatlib_s,kvm,libukofw,$(CONFIG_LIBUKOFW)))
+
+LIBUKOFW_CINCLUDES-y         += -I$(LIBUKOFW_BASE)/include
```
These need to be conditional to `CONFIG_LIBUKOFW`, otherwise they are included even when the library is not enabled, eg:
```
LIBUKOFW_CINCLUDES-$(CONFIG_LIBUKOFW)
```
Similarly for the rest.

michpappas commented on plat/kvm/Makefile.uk:
```diff
@@ -24,6 +23,7 @@ LIBKVMPLAT_ASINCLUDES-y        += -I$(UK_PLAT_COMMON_BASE)/include
 LIBKVMPLAT_CINCLUDES-y         += -I$(LIBKVMPLAT_BASE)/include
 LIBKVMPLAT_CINCLUDES-y         += -I$(UK_PLAT_COMMON_BASE)/include
 LIBKVMPLAT_CINCLUDES-y         += -I$(UK_PLAT_DRIVERS_BASE)/include
+LIBKVMPLAT_CINCLUDES-y         += -I$(CONFIG_UK_BASE)/lib/ukofw/include
```
This one should go away. To make the headers of `libukofw` publicly available what you need to do instead is define the following in `lib/ukofw/Makefile.uk`:
```
ASINCLUDES-$(CONFIG_LIBUKOFW)         += -I$(LIBUKOFW_BASE)/include
CINCLUDES-$(CONFIG_LIBUKOFW)          += -I$(LIBUKOFW_BASE)/include
CXXINCLUDES-$(CONFIG_LIBUKOFW)        += -I$(LIBUKOFW_BASE)/include
```

Notice the diference between `CINCLUDES` (public) and `LIBUKOFW_CINCLUDES` (headers available internally to the library).

michpappas commented on plat/common/arm/lcpu.c:
```diff
@@ -34,11 +34,11 @@
 #include <uk/plat/common/acpi.h>
 #include <uk/plat/lcpu.h>
 #include <uk/plat/io.h>
+#include <uk/ofw/fdt.h>
```
There are more of these that need to be updated:
```bash
bash$ grep -r '<ofw/fdt.h'
plat/drivers/virtio/virtio_mmio.c:#include <ofw/fdt.h>
plat/drivers/gic/gic-v2.c:#include <ofw/fdt.h>
plat/drivers/gic/gic-v3.c:#include <ofw/fdt.h>
plat/drivers/rtc/pl031.c:#include <ofw/fdt.h>
plat/common/pci_ecam.c:#include <ofw/fdt.h>
plat/common/platform_bus.c:#include <ofw/fdt.h>
lib/ukofw/gic_fdt.c:#include <ofw/fdt.h>
lib/ukofw/fdt.c:#include <ofw/fdt.h>
```

michpappas commented on lib/ukofw/Makefile.uk:
```diff
@@ -0,0 +1,8 @@
+$(eval $(call addplatlib_s,kvm,libukofw,$(CONFIG_LIBUKOFW)))
+
+LIBUKOFW_CINCLUDES-y         += -I$(LIBUKOFW_BASE)/include
+LIBUKOFW_CINCLUDES-y         += -I$(CONFIG_UK_BASE)/lib/fdt/include
```
This line is not required, as `lib/fdt` makes its headers publicly available (and hence to this library) via the `CINCLUDES` directives in it's `Makefile.uk`. If you experience compile errors you need to include the necessary headers where these are missing (eg `libfdt.h` in `include/uk/fdt.h`).

michpappas reviewed (changes requested):
@rares-miculescu besides the review comments below, please consider the following:
* Update the include guards in the migrated headers in `lib/ukofw/include/uk`. These should be: `_PLAT_DRIVER_PFW_FDT_H_` -> `__UK_OFW_FDT_H__` and `__PLAT_DRV_GIC_FDT_H__` -> `__UK_OFW_GIC_FDT_H__`. 
* Add an `lib/ukofw/exportsyms.uk` with the public symbols provided by `libukofw`.
* Rebase to `unikraf/staging` and remove the dependency PRs, as they are not really required to migrate the `ofw` library.
* Provide the changes in `plat/` under a separate PR. The general rule is that changes in each library need to come in a separate commit.
* Update your commit message to use assertive voice, ie replace `Moving` with `Move`. The styling of the commit messages has been raised by other reviewers a few times already!
* Remove the `Draft` flag from the PR.

michpappas commented:
One more: I noticed that the source branch is called `move-ofw-test`. In future PRs please prefer branch names that indicate that this is final work.

Commit:
plat: Move ofw from plat/drivers/

Move ofw driver from plat/drivers/ into lib/.
Update files that depend on headers from the driver,
as well as configuration from plat/kvm/.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

Commit:
lib/ukofw: Move ofw into lib/

Move ofw driver from plat/drivers to lib/ukofw.
Create and update configuration files for migration.

Signed-off-by: Rares Miculescu <miculescur@gmail.com>

michpappas commented on lib/ukofw/Makefile.uk:
```diff
@@ -0,0 +1,8 @@
+CXXINCLUDES-$(CONFIG_LIBUKOFW) += -I$(LIBUKOFW_BASE)/include
+
+LIBUKOFW_SRCS-y                += $(LIBUKOFW_BASE)/fdt.c
+LIBUKOFW_SRCS-$(CONFIG_LIBGIC) += $(LIBUKOFW_BASE)/gic_fdt.c
```
Actually this does not really depend on GIC (as in the GIC library), it's just a function that parses some nodes related to GIC in the device-tree, so it can be unconditionally enabled. There are a few other options, but I consider refactoring out of the scope of this PR.

michpappas commented on lib/ukofw/include/uk/ofw/fdt.h:
```diff
@@ -0,0 +1,232 @@
+ */
+uint64_t fdt_translate_address_by_ranges(const void *fdt,
+				int node_offset, const fdt32_t *regs);
+#endif
```
Please add a comment here:
```
endif /* __UK_OFW_FDT_H__ */
```

michpappas commented on lib/ukofw/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBUKOFW
+       bool "ukofw: Device tree helper functions"
+       default n
+       depends on ARCH_ARM_64
```
Should depend on `LIBFDT` instad of `ARCH_ARM_64`, as it uses `libfdt` it internally. Then if `CONFIG_LIBFDT` is not enabled for a given architecture this menu will not be visible. Moreover, `libfdt` is already enabled for any architectures that require it in `plat/kvm/Config.uk` so `ARCH_ARM_64` implicitly applies here too. So:
```
- depends on ARCH_ARM_64
- select LIBFDT
```

michpappas reviewed (changes requested):
Hi @rares-miculescu just a few minor changes I missed earlier. Other than that it looks good 👍🏼

michpappas reviewed (approved):
@rares-miculescu thanks!

There are a few things that could be improved such as the library name that doesn't match the naming in the API, or that the API is not prefixed by `uk_`, but refactoring is out of scope of plat-rearch.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on plat/common/arm/generic_timer.c:
```diff
@@ -36,7 +36,7 @@
 #include <uk/bitops.h>
 #include <uk/plat/common/cpu.h>
-#include <ofw/gic_fdt.h>
+#include <uk/ofw/gic_fdt.h>
```
Thought this should become `uk/<header>.h`?

mogasergiu reviewed (commented):
Just a small question. And then the main question: why are these two commits not squashed into one? So first commit basically removes the `.c`/`.h` files but creates a directory with a `Makefile.uk`/`Config.uk` for them while also updating the header inclusion paths (of headers that for this commit do not exist anymore)? And then the second commit adds the files? i.e. first commit breaks the build, second commit repairs it.

Is there a reason for this? Otherwise I would have all of this in one commit.

mogasergiu reviewed (approved):
All good on my side.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

StefanJum reviewed (commented):
Shouldn't `libofw` depend on `ARCH_ARM_64`?
Everything seems to work fine besides that.

michpappas commented:
> Shouldn't `libofw` depend on `ARCH_ARM_64`? Everything seems to work fine besides that.

`libofw` should depend on `libfdt`, which in turn is enabled for any architecture that requires device-tree.

StefanJum commented:
> > Shouldn't `libofw` depend on `ARCH_ARM_64`? Everything seems to work fine besides that.
> 
> `libofw` should depend on `libfdt`, which in turn is enabled for any architecture that requires device-tree.

Ah, makes sense, thanks. All good on my side too then.

StefanJum reviewed (approved):
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
